File: backend/app/models/mind_map.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import Column, DateTime, ForeignKey, Text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import relationship
from app.db.base import Base


class MindMapLayout(Base):
    """Stores ReactFlow node positions per epic so the map doesn't re-arrange on reload."""
    __tablename__ = "mind_map_layouts"

    id         = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    epic_id    = Column(UUID(as_uuid=True), ForeignKey("epics.id"), nullable=False, unique=True)
    layout     = Column(JSONB, nullable=False, default=dict)   # { "nodeId": { "x": 0, "y": 0 } }
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    epic = relationship("Epic")


class TaskMindMapNote(Base):
    """Intern's approach/logic note for their task node on the epic mind map."""
    __tablename__ = "task_mind_map_notes"

    task_id    = Column(UUID(as_uuid=True), ForeignKey("tasks.id"), primary_key=True)
    note       = Column(Text, nullable=True)
    updated_by = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    task    = relationship("Task", back_populates="mind_map_note")
    updater = relationship("User", foreign_keys=[updated_by])

File: frontend/Spacepoint-interns/src/api/mindmap.ts
import api from "./client"
import type { Epic } from "@/types"

export interface MindMapLayout {
  id: string
  epic_id: string
  layout: Record<string, { x: number; y: number }>
  updated_at: string
}

export interface TaskNote {
  task_id: string
  note: string | null
  updated_at: string
}

// ── Layout ────────────────────────────────────────────────────────────────────

export const getLayoutApi = (epicId: string, role: "admin" | "leader" | "intern") =>
  api.get<MindMapLayout>(`/${role}/epics/${epicId}/mind-map`).then((r) => r.data)

export const saveLayoutApi = (
  epicId: string,
  layout: Record<string, { x: number; y: number }>,
  role: "admin" | "leader"
) => api.patch<MindMapLayout>(`/${role}/epics/${epicId}/mind-map`, { layout }).then((r) => r.data)

// ── Epic fetch ────────────────────────────────────────────────────────────────

export const getEpicForMapApi = (epicId: string, role: "admin" | "leader" | "intern") =>
  api.get<Epic>(`/${role}/epics/${epicId}`).then((r) => r.data)

// ── Task notes ────────────────────────────────────────────────────────────────

export const getTaskNoteApi = (taskId: string, role: "admin" | "leader" | "intern") =>
  api.get<TaskNote>(`/${role}/tasks/${taskId}/mind-map-note`).then((r) => r.data)

export const updateTaskNoteApi = (taskId: string, note: string) =>
  api.patch<TaskNote>(`/intern/tasks/${taskId}/mind-map-note`, { note }).then((r) => r.data)
